### PR TITLE
Tree Attention

### PR DESCRIPTION
**Pull request by hikettei**

### CODE REVIEW

File: lalamo/modules/token_mixers/state/kv_cache.py [hunk 1]
```diff
@@ -13,6 +13,46 @@
...
+def _build_tree_attention_mask(
+    total_capacity: int,
+    prefix_length: Int[Array, ""] | int,
+    parent_indices: Int[Array, " nodes"],
+    has_sinks: bool,
+    max_depth: int,
+) -> Bool[Array, "nodes total_capacity"]:
+    """Tree attention mask: each draft node attends to prefix + ancestors + self."""
+    prefix_length = jnp.asarray(prefix_length, dtype=jnp.int32)
+    num_nodes = parent_indices.shape[0]
```
Comment (knyazer) [inline lalamo/modules/token_mixers/state/kv_cache.py]: prefer tuple unpacking to indexing

File: lalamo/modules/token_mixers/state/kv_cache.py [hunk 2]
```diff
@@ -13,6 +13,46 @@
...
+    n = parent_indices.shape[0]
+    ancestors = jnp.eye(n, dtype=jnp.bool)
+    for _ in range(max_depth):
+        parents_row = ancestors[parent_indices]
+        valid = (parent_indices >= 0)[:, None]
+        ancestors = ancestors | (parents_row & valid)
+    return ancestors
+
+
+def _build_tree_attention_mask(
```
Comment (knyazer) [inline lalamo/modules/token_mixers/state/kv_cache.py]: let's maybe not call this functions _blahblah, just blahblah is good enough; to make sure they are private just don't export them through __all__

Comment (knyazer) [inline lalamo/modules/token_mixers/state/kv_cache.py]: let's wrap this function into eqx.filter_jit explicitly, you use & and | very heavily inside and these ops make DISABLE_JIT=True behave in funny ways (python default behaviour is different from jax default behaviour). I think here its fine since everything you do is boolean, but just to be careful wrapping it into eqx.filter_jit will mean that at least any call of this function will behave the same in/out jitted regions

File: lalamo/modules/token_mixers/state/kv_cache.py [hunk 3]
```diff
@@ -123,6 +170,20 @@ def attention_mask(
...
         return result
 
+    def tree_attention_mask(
+        self,
+        parent_indices: Int[Array, " nodes"],
+        max_depth: int,
+    ) -> Bool[Array, "nodes tokens"]:
+        self._raise_if_batched()
+        total, _, _ = self.keys.shape
+        num_nodes = parent_indices.shape[0]
```
Comment (knyazer) [inline lalamo/modules/token_mixers/state/kv_cache.py]: let's also unpack here maybe? num_nodes, = parent_indices.shape

File: lalamo/modules/token_mixers/state/kv_cache.py [hunk 4]
```diff
@@ -123,6 +170,20 @@ def attention_mask(
...
+        prefix_length = total - num_nodes
+        result = _build_tree_attention_mask(total, prefix_length, parent_indices, self.has_sinks, max_depth)
```
Comment (knyazer) [inline lalamo/modules/token_mixers/state/kv_cache.py]: result is a bad name, "mask" is a better name

File: lalamo/modules/token_mixers/state/kv_cache.py [hunk 5]
```diff
@@ -180,8 +241,19 @@ def attention_mask(
...
+        parent_indices: Int[Array, " nodes"],
+        max_depth: int,
+    ) -> Bool[Array, "nodes tokens"]:
+        self._raise_if_batched()
+        num_nodes = parent_indices.shape[0]
+        prefix_length = self.current_length - num_nodes
+        result = _build_tree_attention_mask(self.capacity, prefix_length, parent_indices, self.has_sinks, max_depth)
+        return result & self.padding_mask[None, :]
+
     @property
```
Comment (knyazer) [inline lalamo/modules/token_mixers/state/kv_cache.py]: isn't this implementation the same as for static kv cache? i think you can just the two and move them into the parent class

File: lalamo/modules/decoder.py [hunk 6]
```diff
@@ -143,6 +143,8 @@ def __call__(
         lengths_without_padding: Int[Array, " batch"] | None = None,
         forward_pass_mode: ForwardPassMode = ForwardPassMode.MULTI_TOKEN,
         forward_pass_config: DecoderForwardPassConfig | None = None,
+        attention_parent_indices: Int[Array, " batch suffix_tokens"] | None = None,
+        attention_max_depth: int | None = None,
```
Comment (knyazer) [inline lalamo/modules/decoder.py]: configs should be the last arguments, attention_parent_indices is good, but attention_max_depth should go into forward_pass_config

File: lalamo/modules/embedding.py [hunk 7]
```diff
@@ -63,6 +63,14 @@ def _prepare_input_weights(self) -> Float[Array, "vocabulary channels"]: ...
...
     def _prepare_output_weights(self) -> Float[Array, "vocabulary channels"]: ...
 
+    @property
+    def embedding_matrix(self) -> Float[Array, "vocabulary channels"]:
+        return self._prepare_input_weights()
+
+    @property
+    def unembedding_matrix(self) -> Float[Array, "vocabulary channels"]:
+        return self._prepare_output_weights()
+
```
Comment (knyazer) [inline lalamo/modules/embedding.py]: you don't actually use these?

File: lalamo/modules/transformer.py [hunk 8]
```diff
@@ -179,6 +179,8 @@ def __call__(
         lengths_without_padding: Int[Array, " batch"] | None,
         forward_pass_mode: ForwardPassMode,
         forward_pass_config: TransformerForwardPassConfig | None,
+        attention_parent_indices: Int[Array, " batch suffix_tokens"] | None = None,
+        attention_max_depth: int | None = None,
```
Comment (knyazer) [inline lalamo/modules/transformer.py]: attention_max_depth goes into forward pass config

File: lalamo/modules/transformer_layer.py [hunk 9]
```diff
@@ -213,6 +213,8 @@ def __call__(
         lengths_without_padding: Int[Array, " batch"] | None = None,
         forward_pass_mode: ForwardPassMode = ForwardPassMode.MULTI_TOKEN,
         forward_pass_config: TransformerLayerForwardPassConfig | None = None,
+        attention_parent_indices: Int[Array, " batch suffix_tokens"] | None = None,
+        attention_max_depth: int | None = None,
```
Comment (knyazer) [inline lalamo/modules/transformer_layer.py]: same

File: lalamo/modules/transformer_layer.py [hunk 10]
```diff
@@ -224,15 +226,31 @@ def __call__(
...
-            length_without_padding=lengths_without_padding,
-        )
+        if attention_parent_indices is not None:
+            batched_mixer_fn = vmap(
+                partial(
+                    self.mixer,
+                    return_updated_state=return_updated_state or return_activation_trace,
+                    attention_max_depth=attention_max_depth,
+                ),
+            )
```
Comment (knyazer) [inline lalamo/modules/transformer_layer.py]: batched_mixed_fn can be moved outside of the condition since its the same

File: lalamo/modules/transformer_layer.py [hunk 11]
```diff
@@ -224,15 +226,31 @@ def __call__(
...
-            state=state,
-            length_without_padding=lengths_without_padding,
-        )
+        if attention_parent_indices is not None:
+            batched_mixer_fn = vmap(
+                partial(
+                    self.mixer,
+                    return_updated_state=return_updated_state or return_activation_trace,
+                    attention_max_depth=attention_max_depth,
+                ),
```
[the remark this answers is not in the document]
Comment (knyazer) [reply lalamo/modules/transformer_layer.py]: btw raising an explicit error would be better, i agree with mr copilot here

File: tests/unit/test_tree_attention.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+            precision=precision,
+        ),
+        transformer_config=transformer_config,
+        vocab_size=vocab_size,
+    )
+    return decoder_config.random_init(key=jax.random.key(4))
+
+
+def test_tree_ancestor_mask_chain() -> None:
+    """Linear chain: each node should attend to itself and all ancestors."""
```
Comment (knyazer) [inline tests/unit/test_tree_attention.py:101]: remove useless comments

Comment (hikettei) [reply tests/unit/test_tree_attention.py:101]: will remove, also i would add my claude.md not to leave extra comments

File: lalamo/modules/transformer_layer.py (hunk 10 above)
Comment (hikettei) [reply lalamo/modules/transformer_layer.py]: agreed, will do

File: lalamo/modules/transformer_layer.py (hunk 9 above)
Comment (hikettei) [reply lalamo/modules/transformer_layer.py]: relocated

File: lalamo/modules/transformer.py (hunk 8 above)
Comment (hikettei) [reply lalamo/modules/transformer.py]: yes

File: lalamo/modules/embedding.py (hunk 7 above)
Comment (hikettei) [reply lalamo/modules/embedding.py]: i have used in #209 but i moved the diff to there

File: lalamo/modules/decoder.py (hunk 6 above)
Comment (hikettei) [reply lalamo/modules/decoder.py]: will do

File: lalamo/modules/token_mixers/state/kv_cache.py (hunk 4 above)
Comment (hikettei) [reply lalamo/modules/token_mixers/state/kv_cache.py]: agreed+renamed

File: lalamo/modules/token_mixers/state/kv_cache.py (hunk 3 above)
Comment (hikettei) [reply lalamo/modules/token_mixers/state/kv_cache.py]: yes

File: lalamo/modules/token_mixers/state/kv_cache.py (hunk 2 above)
Comment (hikettei) [reply lalamo/modules/token_mixers/state/kv_cache.py]: renamed

File: lalamo/modules/token_mixers/state/kv_cache.py (hunk 1 above)
Comment (hikettei) [reply lalamo/modules/token_mixers/state/kv_cache.py]: yes

File: lalamo/modules/transformer_layer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (hikettei) [reply lalamo/modules/transformer_layer.py]: Agreed, I will just throw error for Mamba and DeltaNet.

File: lalamo/modules/token_mixers/state/kv_cache.py (hunk 5 above)
Comment (hikettei) [reply lalamo/modules/token_mixers/state/kv_cache.py]: yes

File: lalamo/modules/token_mixers/state/kv_cache.py (hunk 2 above)
Comment (hikettei) [reply lalamo/modules/token_mixers/state/kv_cache.py]: both helpers wrapped in eqx.filter_jit so &/| behave consistently with JAX_DISABLE_JIT=1 too.

File: lalamo/modules/token_mixers/state/kv_cache.py [hunk 13]
```diff
@@ -13,6 +13,48 @@
...
+    max_depth: int,
+) -> Bool[Array, "nodes nodes"]:
+    """Build ancestor matrix from parent_indices. ancestor[i,j]=True iff j is ancestor of i (or self)."""
+    (num_nodes,) = parent_indices.shape
+    ancestors = jnp.eye(num_nodes, dtype=jnp.bool)
+    for _ in range(max_depth):
+        parents_row = ancestors[parent_indices]
+        valid = (parent_indices >= 0)[:, None]
+        ancestors = ancestors | (parents_row & valid)
+    return ancestors
```
Comment (knyazer) [inline lalamo/modules/token_mixers/state/kv_cache.py]: the problem is that the compilation time will scale linearly here with max_depth, and one can go like "ohh i don't actually know what max depth is, let me pass some large value so we get it for sure" and then it gets compilation 10 minute compilation time/crashes with some random oom error because xla can't handle it or something like that, and its impossible to tell from the outside whether this is an expected behaviour or not.

I didn't mention it in the first round of review because I didn't think about it. The simple solution to this is to just throw up if max depth is more than 30, the correct solution is to avoid max_depth for loop and replcae it with scan, or better yet - if you can figure out how to - avoid this iteration over indices. If you can't figure it out - also fine, i guess, i'll think about it a bit more

Comment (hikettei) [reply lalamo/modules/token_mixers/state/kv_cache.py]: yes thanks for surfacing this. after all i fixed this in eb0065e by dropping max_depth entirely and switching the ancestor closure to a jax.lax.scan that walks nodes once in topological order. (diff's cleaner now too) actually i introduced this parameter in the first round to reduce the compiling time but yeah i should implement this without using for loop.

File: tests/unit/test_tree_attention.py (hunk 12 above)
Comment (knyazer) [reply tests/unit/test_tree_attention.py:101]: ?

File: lalamo/modules/token_mixers/state/kv_cache.py [hunk 14]
```diff
@@ -13,6 +14,57 @@
...
+
+    def step(mask: Bool[Array, "nodes nodes"], i: Int[Array, ""]) -> tuple[Bool[Array, "nodes nodes"], None]:
+        parent = parent_indices[i]
+        parent_row = mask[jnp.maximum(parent, 0)]
+        zero_row = jnp.zeros_like(parent_row)
+        inherited = jnp.where(parent >= 0, parent_row, zero_row)
+        return mask.at[i].set(mask[i] | inherited), None
+
+    mask, _ = jax.lax.scan(step, initial, jnp.arange(num_nodes, dtype=jnp.int32))
+    return mask
```
Comment (knyazer) [inline lalamo/modules/token_mixers/state/kv_cache.py:39]: yeah that's well written, good job

File: lalamo/modules/token_mixers/state/kv_cache.py [hunk 15]
```diff
@@ -44,6 +100,22 @@ def attention_mask(
         sliding_window_size: int | None = None,
     ) -> Bool[Array, "suffix_tokens tokens"]: ...
 
+    def tree_attention_mask(
+        self,
+        prefix_length: Int[Array, ""] | int,
+        parent_indices: Int[Array, " nodes"],
+    ) -> Bool[Array, "nodes tokens"]:
+        self._raise_if_batched()
+        total = self.keys.shape[0]
```
Comment (knyazer) [inline lalamo/modules/token_mixers/state/kv_cache.py]: let's unpack instead of indexing

File: tests/unit/test_tree_attention.py (hunk 12 above)
Comment (hikettei) [reply tests/unit/test_tree_attention.py:101]: i mean some of coding style comments are just me not having internalized the conventions yet sorry for making you flag the same thing multiple times.

File: lalamo/modules/token_mixers/state/kv_cache.py (hunk 15 above)
Comment (hikettei) [reply lalamo/modules/token_mixers/state/kv_cache.py]: done

**Review by knyazer — APPROVED** [review]
lgtm
